iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,603 @@
+    DispatchLazyPush,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Event {
+    Received(Bytes),
+}
+
+/// A message identifier, which is the message content's blake3 hash
+#[derive(Serialize, Deserialize, Clone, Copy, Eq)]
+pub struct MessageId([u8; 32]);
+
+impl MessageId {
+    /// Create a `MessageId` by hashing the message content.
+    ///
+    /// This hashes the input with [`blake3::hash`].
+    pub fn from_content(message: &[u8]) -> Self {
+        Self::from(blake3::hash(message))
+    }
+}
+
+impl From<blake3::Hash> for MessageId {
+    fn from(hash: blake3::Hash) -> Self {
+        Self(hash.into())
+    }
+}
+
+impl std::hash::Hash for MessageId {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.0);
+    }
+}
+
+impl PartialEq for MessageId {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl fmt::Display for MessageId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut text = data_encoding::BASE32_NOPAD.encode(&self.0);
+        text.make_ascii_lowercase();
+        write!(f, "{}", text)
+    }
+}
+impl fmt::Debug for MessageId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut text = data_encoding::BASE32_NOPAD.encode(&self.0);
+        text.make_ascii_lowercase();
+        write!(f, "{}…{}", &text[..5], &text[(text.len() - 2)..])
+    }
+}
+
+#[derive(
+    From, Add, Sub, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug, Hash,
+)]
+pub struct Round(u16);
+
+impl Round {
+    pub fn next(&self) -> Round {
+        Round(self.0 + 1)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub enum Message {
+    /// When receiving Gossip, emit as event and forward full message to eager peer
+    /// and (after a delay) message IDs to lazy peers.
+    Gossip(Gossip),
+    /// When receiving Prune, move the peer from the eager to the lazy set
+    Prune,
+    /// When receiving Graft, move the peer to the eager set and send the full content for the
+    /// included message ID.
+    Graft(Graft),
+    /// When receiving IHave, do nothing initially, and request the messages for the included
+    /// message IDs after some time if they aren't pushed eagerly to us.
+    IHave(Vec<IHave>),
+}
+
+#[derive(Serialize, Deserialize, Clone, derive_more::Debug, PartialEq, Eq)]
+pub struct Gossip {
+    id: MessageId,
+    round: Round,
+    #[debug("<{}b>", content.len())]
+    content: Bytes,
+}
+impl Gossip {
+    pub fn next_round(self) -> Gossip {
+        Gossip {
+            id: self.id,
+            content: self.content,
+            round: self.round.next(),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct IHave {
+    id: MessageId,
+    round: Round,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct Graft {
+    id: Option<MessageId>,
+    round: Round,
```

Comment:
Good point. I rechecked the paper, and there the algorithm when receiving a Graft message is spec'd like this:
```
upon event Receive(GRAFT, mID, round, sender) do
 eagerPushPeers ←− eagerPushPeers ∪ {sender}
 lazyPushPeers ←− lazyPushPeers \ {sender}
 if mID ∈ receivedMsgs do
trigger Send(GOSSIP, sender, m, mID, round, myself)
```
so it uses the `round` of the received graft message as the `round` for the `Gossip` message sent in reply. In my code, the `round` of the `Gossip` message sent in reply to a `Graft` message is as it were when the peer receiving the graft sent that `Gossip` message to its eager peers originally.

I'm trying to make sense of what is better or how it would affect further proceedings, but did not come to a conclusion yet.

Reply:
reading it again I have no idea what's the point of sending this to begin with since the information is redundant and redundancy is not verified. Just for the sake of simplicity I would remove it.

For completeness though, I'm going to take a wild guess. I think maybe the authors wanted to have some kind of "conditional" graft. Something along the lines of: graft only if you have this with less rounds than `r` and then decided to move the logic to the sender side with the `IHave`s to avoid simultaneous grafts. This is based on absolutely nothing, like I said .. _wild_ guess